discoverd/client/client.go

```diff
@@ -1,261 +1,192 @@
 package discoverd
 
 import (
-	"encoding/json"
 	"errors"
 	"fmt"
 	"net/http"
 	"os"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/flynn/flynn/pkg/dialer"
 	"github.com/flynn/flynn/pkg/httpclient"
-	hh "github.com/flynn/flynn/pkg/httphelper"
 	"github.com/flynn/flynn/pkg/stream"
 )
 
-type Service interface {
-	Leader() (*Instance, error)
-	Instances() ([]*Instance, error)
-	Addrs() ([]string, error)
-	Leaders(chan *Instance) (stream.Stream, error)
-	Watch(events chan *Event) (stream.Stream, error)
-	GetMeta() (*ServiceMeta, error)
-	SetMeta(*ServiceMeta) error
-	SetLeader(string) error
-}
-
 var ErrTimedOut = errors.New("discoverd: timed out waiting for instances")
+var PinTTL = 60 * time.Second
+
+type Config struct {
+	Endpoints []string
+}
 
 type Client struct {
-	c *httpclient.Client
+	servers    []*httpclient.Client
+	pinned     int
+	pinUpdated time.Time
+	leader     int
+	mu         sync.RWMutex
 }
 
-func NewClient() *Client {
-	url := os.Getenv("DISCOVERD")
-	if url == "" {
-		url = "http://127.0.0.1:1111"
+func NewClientWithConfig(config Config) *Client {
+	client := Client{}
+	for _, e := range config.Endpoints {
+		client.servers = append(client.servers, client.httpClient(e))
 	}
-	return NewClientWithURL(url)
+	return &client
 }
 
 func NewClientWithURL(url string) *Client {
-	if !strings.HasPrefix(url, "http") {
-		url = "http://" + url
-	}
-	return &Client{
-		c: &httpclient.Client{
-			URL: url,
-			HTTP: &http.Client{
-				Transport:     &http.Transport{Dial: dialer.Retry.Dial},
-				CheckRedirect: redirectPreserveHeaders,
-			},
-		},
-	}
+	return NewClientWithConfig(Config{Endpoints: formatURLs([]string{url})})
 }
 
-func NewClientWithHTTP(url string, hc *http.Client) *Client {
-	if url == "" {
-		url = "http://127.0.0.1:1111"
-	}
-	return &Client{
-		c: &httpclient.Client{
-			URL:  url,
-			HTTP: hc,
-		},
-	}
+func NewClient() *Client {
+	return NewClientWithConfig(defaultConfig())
 }
 
-func redirectPreserveHeaders(req *http.Request, via []*http.Request) error {
-	if len(via) >= 10 {
-		return fmt.Errorf("too many redirects")
-	}
-	if len(via) == 0 {
-		return nil
-	}
-	for attr, val := range via[0].Header {
-		if _, ok := req.Header[attr]; !ok {
-			req.Header[attr] = val
-		}
+func defaultConfig() Config {
+	urls := os.Getenv("DISCOVERD")
+	if urls == "" || urls == "none" {
+		urls = "http://127.0.0.1:1111"
 	}
-	return nil
+	return Config{Endpoints: formatURLs(strings.Split(urls, ","))}
 }
 
-func (c *Client) Ping() error {
-	return c.c.Get("/ping", nil)
+func formatURLs(urls []string) []string {
+	formatted := make([]string, len(urls))
+	for _, u := range urls {
+		if !strings.HasPrefix(u, "http") {
+			u = "http://" + u
+		}
+		formatted = append(formatted, u)
```

Comment:
Given you have created a slice with a non-zero length, this either needs to index into that slice (i.e. `formatted[i] = u`), or just create the slice with a capacity but zero length.